src/Castle.Core/DynamicProxy/AbstractInvocation.cs

```diff
@@ -188,5 +193,31 @@ private MethodInfo EnsureClosedMethod(MethodInfo method)
 			}
 			return method;
 		}
+
+		private sealed class ProceedInfo : IInvocationProceedInfo
+		{
+			private readonly AbstractInvocation invocation;
+			private readonly int interceptorIndex;
+
+			public ProceedInfo(AbstractInvocation invocation)
+			{
+				this.invocation = invocation;
+				this.interceptorIndex = invocation.currentInterceptorIndex;
+			}
+
+			public void Invoke()
+			{
+				var previousInterceptorIndex = invocation.currentInterceptorIndex;
+				try
+				{
+					invocation.currentInterceptorIndex = interceptorIndex;
+					invocation.Proceed();
```

Comment:
At this point the invocation really is complete because an interceptor chose to set a `ReturnValue` and decided not to proceed anymore returning control and completing the method invocation. Obviously in the async scenario we have to deal with the slightly leaky abstraction to handle the async machinery.

What do you think of adding a `completed` (or similar) boolean field to `AbstractInvocation` to enforce a readonly `ReturnValue` and prevent `SetArgumentValue` calls (unfortunately can't do much with the arguments array) after the invocation has completed. This should help catch broken interceptors because it would be a fairly easy mistake to make.

Reply:
> What do you think of adding a `completed` (or similar) boolean field to `AbstractInvocation` to enforce a readonly `ReturnValue` and prevent `SetArgumentValue` calls [...] after the invocation has completed.

For simplicity's sake, let's ignore setting by-ref arguments and focus on setting the return value, which is the more important and common scenario (by-ref parameters don't play well with async).

In an async interception scenario, it's true that setting `ReturnValue` only has an effect on the caller before the first `await`. But it's still useful that subsequent interceptors or the proxy target object can set `ReturnValue` again, because right now, they have no other way of communicating their return value to earlier interceptors.

Consider a case where we intercept a method `Task<int> ComputeAsync()`. We have one interceptor, which (for argument's sake) needs to `await` something ("X"), then wants to `Proceed` to the proxy target object, then use the return value produced by that to compute its own modified return value. This might be implemented as follows:

1. The interceptor sets `ReturnValue` to a non-completed `Task<int>` and `await`s X. This causes the task to be returned to the caller.

2. Now the interceptor `Proceed`s to the proxy target object.

3. The method on the proxy target object is just a regular method that *has to* `return` a `Task<int>` (except if it threw). The invocation object generated by DP will turn that return value into an assignment to `ReturnValue`. (This new return value will never be seen by the original caller.)

4. Now we're back in our interceptor, which can inspect this new `ReturnValue`, `await` it, extract its result, derive its own return value from that, and complete its own `Task<int>` to the final return value.

Now, if we forbade repeated setting of `ReturnValue` by means of a `completed` flag, how would the interceptor be able to receive the return value produced by the proxy target object?

I think we should allow setting `ReturnValue` after interception has completed. But we should perhaps provide documentation that explains how `IInvocation` objects can essentially go "stale" in async scenarios, and that it's an async interceptor's responsibility to restore an `IInvocation` to a clean state (i.e. it should finally restore `ReturnValue` to the value that the caller received).

Shall I try to cook up some documentation?

Reply:
> because right now, they have no other way of communicating their return value to earlier interceptors.

Ah yer, silly me. Yep, reusing `ReturnValue` (and arguments) for the real invocation and continuations is a good compromise, noting this in documentation maybe even with a rough diagram would greatly help users when they have to come here.

> I think we should allow setting `ReturnValue` after interception has completed. But we should perhaps provide documentation that explains how `IInvocation` objects can essentially go "stale" in async scenarios, and that it's an async interceptor's responsibility to restore an `IInvocation` to a clean state (i.e. it should finally restore `ReturnValue` to the value that the caller received).
> 
> Shall I try to cook up some documentation?

Cooking up some documentation definitely sounds like the next step. Talking about the lifetime of an `IInvocation` and how you can use the instance after the fact via continuations is the key.

Reply:
I've made an attempt to documentation in https://github.com/castleproject/Core/pull/439/commits/b378594c0c94cd2befb90eaa329eb5941b04e973. See it rendered at https://github.com/stakx/Castle.Core/blob/invocationproceedinfo/docs/dynamicproxy-async-interception.md.

No nice illustrations for now, as it's getting a little late; but perhaps it's "good enough" for a start? Let me know if you have any requests or suggestions to improve it.